frb_codegen/src/method_utils.rs

```diff
@@ -0,0 +1,181 @@
+    }
+
+    // Returns the name of the struct that this method is for
+    pub fn static_method_return_struct_name(s: &str) -> String {
+        s.split(STATIC_METHOD_MARKER).last().unwrap().to_string()
+    }
+
+    // Returns the name of method itself
+    pub fn static_method_return_method_name(s: &str) -> String {
+        s.split(STATIC_METHOD_MARKER).next().unwrap().to_string()
+    }
+
+    // Returns the name of the struct that this method is for
+    pub fn non_static_method_return_struct_name(s: &str) -> String {
+        s.split(METHOD_MARKER).last().unwrap().to_string()
+    }
+
+    // Returns the name of method itself
+    pub fn non_static_method_return_method_name(s: &str) -> String {
+        s.split(METHOD_MARKER).next().unwrap().to_string()
+    }
+
+    // Tests if a given struct has methods, that is, if the `ir_file` contains
+    // a function that receives the struct as first argument
+    pub fn struct_has_methods(file: &IrFile, the_struct: Option<&IrType>) -> bool {
+        let the_struct = if let Some(s) = the_struct {
+            s
+        } else {
+            return false;
+        };
+        let struct_name = if let StructRef(IrTypeStructRef { name, freezed: _ }) = the_struct {
+            name
+        } else {
+            return false;
+        };
+
+        file.funcs.iter().any(|f| {
+            if let Some(IrField { name, .. }) = f.inputs.get(0) {
+                name.raw.to_case(Case::UpperCamel) == *struct_name
+            } else {
+                false
+            }
+        })
+    }
+
+    fn mark_as_static_method(s: &str, struct_name: &str) -> String {
+        format!("{}{}{}", s, STATIC_METHOD_MARKER, struct_name)
+    }
+
+    fn mark_as_non_static_method(s: &str, struct_name: &str) -> String {
+        format!("{}{}{}", s, METHOD_MARKER, struct_name)
+    }
+}
+
+#[derive(Debug)]
+pub enum MethodInfo {
+    Not,
+    Static { struct_name: String },
+    NonStatic { struct_name: String },
+}
+
+#[derive(Debug)]
+pub struct FunctionName {
+    actual_name: String,
+    method_info: MethodInfo,
+}
+
+impl FunctionName {
+    pub fn new(name: &str, method_info: MethodInfo) -> FunctionName {
+        FunctionName {
+            actual_name: name.to_string(),
+            method_info,
+        }
+    }
+    // assemble it into a string, e.g. `method_name__static_method___TheStructName` etc
+    pub fn serialize(&self) -> String {
+        match &self.method_info {
+            MethodInfo::Not => self.actual_name.clone(),
+            MethodInfo::Static { struct_name } => {
+                MethodNamingUtil::mark_as_static_method(&self.actual_name, struct_name)
+            }
+            MethodInfo::NonStatic { struct_name } => {
+                MethodNamingUtil::mark_as_non_static_method(&self.actual_name, struct_name)
+            }
+        }
+    }
+
+    pub fn deserialize(s: &str) -> Self {
+        let actual_name: String;
+        let method_info = {
```

Comment:
what about:

```
if ... {
  FunctionName {actualName: ..., method_info: ...}
} else {
  FunctionName {...}
}
```

In other words, no need to temporary variable actual_name and method_info

And, use early return seems good

```
if staticmethod { return ... }
if nonstatic { return ...}
return ... 
```